### ISSUE TITLE
quick review

### ISSUE THREAD

**Issue by kakaroto** [zeel01/input-expressions#1]
standard disclaimer of "this isn't a critique of you or your skills, feel free to ignore me, blablabla, etc..."
- in `init` hook, roll function
```js
let str = typeof args == "object"
			? args.reduce((s, a) => s + a.toString().replace(/\s/g, ''), "")
			: args;
```
I'd use `args instanceof Array ? args.reduce...` since anything can be an "object" including `{}` but they won't necessarily have a reduce function. Actually `args.reduce?((s, a) ... ) || args` might even work better (for Array like objects that aren't arrays).
Also, wouldn't it work the same if you did `args.join("").replace(/\s/g, "")` instead of the reduce (but increased code readability)? 

- in other hooks
```js
        inputs.off("change");
	inputs.on("change", ...
```
personal preference, you can do `inputs.off("change").on("change", ...)` as they are chainable.
- in hooks
`let current = getProperty(sheet.actor.data, event.target.name);` can use `const` since you don't change `current`
- in hooks
```js
_onChangeInputDelta.bind({
			current,
			entity: sheet.actor.data
		})(event);
```
You shouldn't do a `bind` here (which creates a new bound function) if it's just to call it, instead, you should just `call` or `apply` (I never remember which is the right one) and do `_onChangeInputDelta.call({}, event)` directly
- Same
I'm not sure if using that object with 2 fields as `this` makes sense, why isn't it just a regular argument to the function, and you call it without a `.call` or `.bind` ?

- renderActorSheet/renderItemSheet
These two are essentially the same, duplicated code and risk of fixing a bug in one but not in the other makes me suggest extracting that function on its own and changing the hook into : 
```js
renderHook = function (element, data) {
	element.find('[data-dtype="Number"]')
            .off("change")
	    .on("change", (event) => {
		const current = getProperty(data, event.target.name);
		_onChangeInputDelta.call({
			current,
			entity: data
		}, event);
	});
}
Hooks.on("renderItemSheet", (sheet, element, data) => renderHook(element, sheet.item.data));
Hooks.on("renderActorSheet", (sheet, element, data) => renderHook(element, sheet.actor.data));
```

- renderTokenHUD
  - same as above about chaining the calls and the bind issue
  - the `bar` variable isn't modified, so should be `const` instead of `let`
  - target is only used inside the `if (bar)` but gets defined outside of it
  - the `const current = c` just to pass current as argument is unnecessary, you can do `current: c` though I think it would make more sense to simply rename the `c` variable into `current` directly
You do this : 
```js
			c = getProperty(hud.object.actor.data.data, target);
			if (typeof c == "object") {
				target += ".value";
				c = getProperty(hud.object.actor.data.data, target);
			}
```
which could be rewritten in a more readable way like this : 
```js
			c = getProperty(hud.object.actor.data.data, target);
			if (typeof c == "object") {
				c = c.value;
			}
```
(would also probably suggest a comment above the if with regards to the `.value` attributes in dnd5e so 6 months from now, it still makes sense)
  - With the above change, `target` is now used once and not modified, and only within the `if (bar)` clause, so can become a const

Result becomes : 
```js
		const bar = input.dataset.bar;
		let current;
		if (bar) {
			const target = hud.object.data[bar].attribute;
			current = getProperty(hud.object.actor.data.data, target);
			if (typeof c == "object") {
				current = current.value;
			}
		}
		else {
			current = hud.object.data[input.name];
		}
		_onChangeInputDelta({
			current,
			entity: hud.object.actor.data
		}, event);
		hud._onAttributeUpdate(event);
```

- I would put the _onChangeInputDelta after all the hooks, as I had to skip it because I didn't know yet where it got used. By putting it at the end of the file, it makes for a more linear progression with the functions used first being at the top and what they call being lower.

- The use of `this` is weird imho. Instead of : 
```js
function _onChangeInputDelta(event) {
	// Helpers
	
	let entity = this.entity;
	let data = this.entity.data;
        // etc..
```
I would do : 
```
function _onChangeInputDelta({current, entity}, event) {
	// Helpers
	
	let data = entity.data;
        // etc..
```
This also of course avoids the whole `bind/call/apply` stuff.

- The data, attributes, abilities, scope variables aren't modified, so should be `const`
- You set up the scope, then check for input.value being "" and return. I'd do the check earlier so you don't setup the scope if you're not going to use it.
- in the catch, you do `ui.notifications.error(e);` I'm not sure if it makes a difference or if it's the same thing as doing `ui.notifications.error(e.message || e);`, that's what I've always been using when I needed a string out of the catch, but maybe it's not necessary.
- left over commented line that used the `eval` can probably removed at this point, so it doesn't pollute the code

Overall, it all looks good, just my nitpicks with regards to style and cleaner code for maintainability, but I didn't notice anything that would lead to a bug anyway (other than the roll args being a non array-like object, though not sure where that situation could happen).  :)

**Comment by zeel01** [comment]
Haha, thanks. Most of this is a consequence of the development process and adapting code that worked one way elsewhere (in Monster Blocks) and making it work here.

It absolutely needs heavily refactored though, as you have pointed out there are a lot of vestigial things in there. 

One point to contend though: I don't like to define something after I use it. Yes, it's allowed in Javascript, but I guess too much C++ has gotten into me and I prefer to define before use.

**Comment by zeel01** [comment]
Oh, `this` is a leftover from my initial implementation to use `eval()`, in Monster Blocks. I liked that `this` was available to the code, and continued to emulate it on this module. However, by dropping `eval` for Math.js, it's no longer of any use.

**Comment by kakaroto** [comment]
Yeah, I could see the evolution of the code and vestigial constructs :) Good point about `this` for the content within the eval, that makes sense, though it's unneeded now with the use of math.js.
I also hate not defining every variable at the start of the function, I'm getting slowly used to it with js now. To me, it wasn't C++ but rather working for a while on a system that only supported C89 (which didn't allow variables being declared after the first statement).
